subprojects/scala/src/main/java/org/gradle/api/tasks/scala/internal/GenerateScaladoc.java

```diff
@@ -69,9 +71,25 @@ public void execute() {
     }
 
     private void invokeScalaDoc(List<String> args) throws ClassNotFoundException, NoSuchMethodException, InstantiationException, IllegalAccessException, InvocationTargetException {
-        Class<?> scaladocClass = Thread.currentThread().getContextClassLoader().loadClass("scala.tools.nsc.ScalaDoc");
-        Method process = scaladocClass.getMethod("process", String[].class);
+        ClassLoader scalaClassLoader = Thread.currentThread().getContextClassLoader();
+        String scalaVersion = getScalaVersion(scalaClassLoader);
+
+        String scaladocFqName = scalaVersion.charAt(0) == '3' ? "dotty.tools.scaladoc.Main" : "scala.tools.nsc.ScalaDoc";
+        String scaladocEntryName = scalaVersion.charAt(0) == '3' ? "run" : "process";
+
+        Class<?> scaladocClass = Thread.currentThread().getContextClassLoader().loadClass(scaladocFqName);
+        Method process = scaladocClass.getMethod(scaladocEntryName, String[].class);
         Object scaladoc = scaladocClass.getDeclaredConstructor().newInstance();
         process.invoke(scaladoc, new Object[]{args.toArray(new String[0])});
     }
+
+    private static String getScalaVersion(ClassLoader scalaClassLoader) {
```

Comment:
I wonder if we could not inject that information in the worker instead of having to load something here. After all, we had to ask the same question earlier.

Reply:
I think this is impossible since the plugin is universal (it's just `scala`) and we determine the version at runtime by the scala library dependency, but maybe you have some idea how to workaround it.

Reply:
The path would be something like: `ScalaBasePlugin` -> `ScalaDoc` -> `ScaladocParameters` -> `GenerateScaladoc`

makes sense?